src/render.rs

```diff
@@ -113,320 +133,291 @@ impl Render {
+    fn bar_buffer_offset(&self) -> usize {
+        if self.bars.is_some() {
+            self.theme.width() * self.theme.bar_height()
+        } else {
+            0
         }
     }
 
-    let height = 40.0;
-    let padding = 10.0;
-    let scale = Scale {
-        x: height,
-        y: height,
-    };
-
-    let v_metrics = font.v_metrics(scale);
-    let glyphs = font.layout(
-        player_name,
-        scale,
-        rusttype::point(padding, padding + v_metrics.ascent),
-    );
-
-    for g in glyphs {
-        if let Some(bb) = g.pixel_bounding_box() {
-            // Poor man's anti-aliasing.
-            g.draw(|left, top, intensity| {
-                let left = left as i32 + bb.min.x;
-                let top = top as i32 + bb.min.y;
-                if 0 <= left
-                    && left < theme.width() as i32
-                    && 0 <= top
-                    && top < theme.bar_height() as i32
-                    && intensity >= 0.01
-                {
-                    if intensity < 0.5 && text_color == theme.text_color() {
-                        view[(top as usize, left as usize)] = theme.med_text_color();
-                    } else {
-                        view[(top as usize, left as usize)] = text_color;
-                    }
-                }
-            });
+    /// Returns the number of pixels required to render one player bar.
+    fn bar_pixel_offset(&self) -> usize {
+        if self.bars.is_some() {
+            self.theme.bar_height()
         } else {
-            text_color = theme.text_color();
+            0
         }
     }
+
+    /// Returns a tuple of (image height, image width)
+    fn image_dims(&self) -> (usize, usize) {
+        Render::get_image_dims(self.theme, self.bars.is_some())
+    }
+
+    fn get_image_dims(theme: &Theme, bars: bool) -> (usize, usize) {
+        (theme.width(), theme.height(bars))
+    }
+}
+
+impl Iterator for Render {
+    type Item = BytesMut;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.state {
+            RenderState::Preamble => {
+                self.encoder
+                    .set_repeat(Repeat::Infinite)
+                    .expect("encode repeat");
+                self.render_comment();
+                self.render_player_bar(PlayerBar::Top);
+                self.render_player_bar(PlayerBar::Bottom);
+                self.render_frame(0);
+            }
+            RenderState::Frame(frame_index) => {
+                if frame_index < self.frames.len() {
+                    self.render_frame(frame_index);
+                } else {
+                    self.render_last_frame();
+                }
+            }
+            RenderState::Complete => return None,
+        };
+
+        // Return the bytes rendered during this frame. The `split_off` function resets the encoder
+        // buffer to have length 0 again
+        Some(self.encoder.get_mut().get_mut().split_off(0))
```

Comment:
Nice trick! I thought getting individual frames out of `gif` would not be so easy.